src/mark.typ

```diff
@@ -102,55 +103,114 @@
-        if slant not in (none, 0%) {
-          if type(slant) == ratio {
-            slant /= 100%
-          }
-          matrix.transform-shear-x(slant)
-        },
-        if flip or reverse {
-          matrix.transform-scale({
-            if flip {
-              (y: -1)
-            }
-            if reverse {
-              (x: -1)
-            }
-          })
+  assert(style.anchor in ("tip", "base", "center"))
+  let tip = mark.tip
+  let base = mark.base
+  let origin = mark.at(style.anchor)
+
+  mark.offset = vector.dist(origin, tip)
+
+  let t = (
+    /* Translate & rotate to the target coordinate & direction */
+    matrix.transform-translate(..pos),
+    matrix.transform-rotate-dir(dir, up),
+    matrix.transform-rotate-z(-90deg),
+
+    /* Apply mark transformations */
+    if reverse {
+      matrix.transform-translate(-mark.length, 0, 0)
+    },
+    if slant not in (none, 0%) {
+      if type(slant) == ratio {
+        slant /= 100%
+      }
+      matrix.transform-shear-x(slant)
+    },
+    if flip or reverse {
+      matrix.transform-scale({
+        if flip {
+          (y: -1)
         }
-      ).filter(e => e != none)
-    ),
+        if reverse {
+          (x: -1)
+        }
+      })
+    },
+
+    /* Rotate mark to have base->tip on the x-axis */
+    matrix.transform-rotate-z(vector.angle2(base, tip)),
+
+    /* Translate mark to have its anchor (tip, base, center) at (0,0) */
+    matrix.transform-translate(..vector.scale(origin, -1)),
+  )
+
+  mark.drawables = drawable.apply-transform(
+    matrix.mul-mat(..t.filter(m => m != none)),
     mark.drawables
   )
+
   return mark
 }
 
+#let _eval-mark(ctx, mark, style) = {
+  if "eval-mark-guard" in ctx {
+    panic("Recursive mark drawing is not allowed")
+  }
+  ctx.eval-mark-guard = true
+
+  ctx.groups = ()
+  ctx.nodes = (:)
+  ctx.transform = matrix.ident()
+
+  import "/src/draw.typ"
+  let body = draw.group({
+    draw.set-style(
+      stroke: style.at("stroke", default: none),
+      fill: style.at("fill", default: none),
+      mark: none,
+      line: (mark: none),
+      bezier: (mark: none),
+      arc: (mark: none),
+    )
+    mark
+  }, name: "mark")
+  let (ctx: ctx, bounds: bounds, drawables: drawables) = process.many(ctx, body)
+  let anchor-fn = ctx.nodes.at("mark").anchors
+
+  // Check if the mark has named anchor
+  let has-anchor(name) = {
+    return name in (anchor-fn)(())
+  }
+
+  // Fetch special mark anchors
+  let get-anchor(name, default: none) = {
+    if default != none {
+      if not has-anchor(name) {
+        return default
+      }
+    }
+    return (anchor-fn)(name)
+  }
+
+  let tip = get-anchor("tip", default: (0, 0, 0))
+  let base = get-anchor("base", default: tip)
+  let center = get-anchor("center", default: vector.lerp(tip, base, .5))
+
+  return (
+    tip: tip,
+    base: base,
+    center: center,
+    length: vector.dist(tip, base),
+    drawables: drawables,
+  )
+}
+
```

Comment:
It took me a while to figure out exactly what this function does as it does several things at once. I'm not happy with the quality of this code but I don't know right now how to improve it...

Reply:
This is not a helpful comment. I renamed the function to better reflect what its purpose is (rendering a mark and finding the marks special anchors).